Add completion for %colors magic

tomowind opened the pull request:
Fixes #8901

tomowind committed:
Add completion for %colors magic

takluyver commented on IPython/core/completer.py:
```diff
@@ -1879,6 +1891,14 @@ def _complete(self, *, cursor_line, cursor_pos, line_buffer=None, text=None,
 
+        # Do magic arg matches
+        for matcher in self.magic_arg_matchers:
+            matches = [(m, matcher.__qualname__) for m in matcher(line_buffer)]
```
I don't think we need to put `matcher.__qualname__` into the list - it's going to be the same for every item anyway.

Carreau replied:
You actually get two of these: 
```
self.magic_arg_matchers = [
            self.magic_config_matches,
            self.magic_color_matches,
        ]
```

Origin is not super important but useful for debug, and to deduplicate completions in a smart ways. It's cheap enough o keep.

takluyver replied:
Right, but you never mix entries from the two matchers. You could achieve the same thing like this:

```python
for matcher in self.magic_arg_matchers:
    matches = matcher(line_buffer)
    if matches:
        origins = [matcher.__qualname__] * len(matches)
        return text, matches, origins, ()
```

i.e. skip putting the origins into a list-of-tuples and then separating them back out.

takluyver commented on IPython/core/completer.py:
```diff
@@ -1879,6 +1891,14 @@ def _complete(self, *, cursor_line, cursor_pos, line_buffer=None, text=None,
+            if matches:
+                matches2 = [m[0] for m in matches]
+                origins = [m[1] for m in matches]
+                return text, matches2, origins, {}
```
It looks like the fourth return value below is a list or a tuple, not a dict.

Carreau replied:
Hum, I missed that it can return 4 things, I'l have to dig into that. The type annotation is wrong on this function for sure then.

Carreau commented on IPython/core/completer.py:
```diff
@@ -1182,13 +1186,11 @@ def magic_matches(self, text):
         return [cast_unicode_py2(c) for c in comp]
 
-    def magic_config_matches(self, text):
+    def magic_config_matches(self, line_buffer):
```
nitpicking but changing a positional parameter name, is a change of API as it can be used a a keyword argument.

Carreau replied:
Adding on that, but because it's new and has not been released it ok :-)  Just noting it.

Carreau committed:
Minimal fix:

line_buffer is an old denomination, it can be multiline text. so update
names in new functions.

Fix types annotations.

Carreau commented:
Rebased and fix a couple of return type where `{}` was a set. I also added a couple of type annotations at the same time.

tomowind commented:
Thanks. Let me pull the changes and revisit this one.

Carreau commented:
This work, I'm going to merge ahead, and we can still make subsequent PRs if necessary.
